fatrace.c

```diff
@@ -225,15 +228,73 @@ show_pid (pid_t pid)
     return true;
 }
 
+static bool
+is_valid_string(const char *str)
+{
+    mbstate_t st = {0};
+    char32_t    cp;
+    while (1) {
+        size_t n = mbrtoc32(&cp, str, MB_CUR_MAX, &st);
+        if (n == (size_t)-1 || n == (size_t)-2) /* illegal or incomplete */
+            return false;
+        if (n == 0) /* reached terminating NUL */
+            return true;
+        str += n; /* advance to next glyph */
+    }
+}
+
+static bool
+print_json_str (const char* key, const char* value) {
+    bool valid_value = is_valid_string(value);
+    if (valid_value) {
+        printf("\"%s\":\"", key);
+        for (int i = 0; value[i] != 0; i++) {
+            unsigned char c = value[i];
+            switch(c) {
+                case '"':  printf("\\\""); continue;
+                case '\\': printf("\\\\"); continue;
+                case '\b': printf("\\b"); continue;
+                case '\f': printf("\\f"); continue;
+                case '\n': printf("\\n"); continue;
+                case '\r': printf("\\r"); continue;
+                case '\t': printf("\\t"); continue;
+            }
+            if (c < 0x20 || c == 0x7e) {
+                printf("\\u%04x", c); continue;
+            }
+            putchar(c);
+        }
+        putchar('"');
+    } else {
+        printf("\"%s_raw\":[", key);
+        for (int i = 0; value[i] != 0; i++)
+            printf(i ? ",%d" : "%d", (unsigned int)(unsigned char)(value[i]));
+        putchar(']');
+    }
+    return !valid_value;
+}
+
 /**
  * print_event:
  *
  * Print data from fanotify_event_metadata struct to stdout.
  */
+#define MAX_PROBLEMS 1000 /* Keep MAX_PROBLEMS greater than or equal to the
+                             number of calls to add_problem, and make sure each
+                             call to add_problem() can be triggered no more than
+                             once. */
+#define add_problem(problem_str) \
+    do { \
+        problems[problem_idx++] = problem_str; \
```

Comment:
`problem` and `problem_idx` are dangling here, it's neither a global variable nor a macro argument. I see why you did it, but this is type-unsafe and brittle. I realize that making it a function which gets `char **problems, int *problem_idx, const char* message)` is less convenient to use. So grudgingly accepting this. (C is just terrible :grin: )